app/get-exchange-rate/getExchangeRateInteractor.js

```diff
@@ -0,0 +1,69 @@
+const ValidationError = require('../errors/validationError');
+
+class GetExchangeRateInteractor {
+    constructor({
+        presenter,
+        validator,
+        exchangerFactory,
+        exchangeRateRepository,
+        getExchangeRateResponseBuilder,
+        loggerProvider,
+        exchangeRateProvider,
+        dateProvider
+    }) {
+        this.presenter = presenter;
+        this.validator = validator;
+        this.exchangerFactory = exchangerFactory;
+        this.exchangeRateRepository = exchangeRateRepository;
+        this.responseBuilder = getExchangeRateResponseBuilder;
+        this.logger = loggerProvider.create(GetExchangeRateInteractor.name);
+        this.exchangeRateProvider = exchangeRateProvider;
+        this.dateProvider = dateProvider;
+    }
+
+    async execute(request) {
+        const errors = this.validator.validate(request);
+
+        if (errors.length > 0) {
+            this.presenter.presentFailure(new ValidationError(errors));
+            return;
+        }
+
+        const rates = await this.exchangeRateRepository.getRateByDate(request.date);
+        if (!rates) {
+            // probably currency api would work if i request actual rates as historical with current date
+            // but let it be that way for now
+            let rates;
+            if (this.dateProvider.getCurrentDate() === request.date) {
```

Comment:
If the only purpose of DateProvider to convert date to some format, that should be later be compared... than u can use defualt JS Date object)
```suggestion
            if (new Date() === new Date(request.date)) {
```